plugin/manager.go

```diff
@@ -0,0 +1,185 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	hplugin "github.com/hashicorp/go-plugin"
+)
+
+// PluginManager can start new plugins watch and kill all plugins.
+type PluginManager struct {
+	Interval time.Duration
+
+	sources     []withClient[Source]
+	destination []withClient[Destination]
+	m           sync.Mutex
+}
+
+// NewDestination returns a new Destination interface from a plugin path and configuration.
+func (pm *PluginManager) NewDestination(path string, config map[string]any) (Destination, error) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	c := client(path)
+	cp, err := c.Client()
+	if err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to create rpc client interface: %w", err)
+	}
+	d, err := newDestination(cp)
+	if err != nil {
+		c.Kill()
+		return nil, err
+	}
+
+	if err := d.Configure(config); err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to configure destination: %w", err)
+	}
+
+	pm.destination = append(pm.destination, withClient[Destination]{t: d, c: c, path: path, config: config})
+
+	return d, nil
+}
+
+// NewSource returns a new Source interface from a plugin path and configuration.
+func (pm *PluginManager) NewSource(path string, config map[string]any) (Source, error) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	c := client(path)
+	cp, err := c.Client()
+	if err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to create rpc client interface: %w", err)
+	}
+	s, err := newSource(cp)
+	if err != nil {
+		c.Kill()
+		return nil, err
+	}
+	if err := s.Configure(config); err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to configure source: %w", err)
+	}
+	pm.sources = append(pm.sources, withClient[Source]{t: s, c: c, path: path, config: config})
+	return s, nil
+}
+
+// Stop will block until all rpc clients are closed.
+// After Stop was called the PluginManager should not be used anymore.
```

Comment:
I wonder if we want to have an internal field called `stopped boolean` and once the PM is stopped, any subsequent calls to methods should error?
Otherwise, by looking at the code I don't see anything that prevents a PM from being reused after a stop. The only state, sources and destinations, are empty slices after the call so it seems fine?

Reply:
yes go-plugin does something similar but with a global variable:
```go
// This must only be called _once_.
func CleanupClients() {
	// Set the killed to true so that we don't get unexpected panics
	atomic.StoreUint32(&Killed, 1)
...

```

Reply:
In our case it seems safe to call `Stop()` as many times as we like and to even add new plugins afterwards, no?

Reply:
yes the comment can be removed now because we free all old plugins. It should be clear though the it deletes the old ones and they can't be reused